docs: add section to upgrade guide on lazy loading AngularJS (#30541)
angular/angular: aio/content/guide/upgrade.md

<code-example path="upgrade-module/src/app/a-to-ajs-providers/hero-detail.component.ts" header="hero-detail.component.ts">
</code-example>

## Lazy Loading AngularJS

When migrating large applications from AngularJS to Angular using a hybrid approach, you want to migrate some of the most commonly used features first. For example, a YouTube client application might migrate the home page, then the player page, then search, and finally the settings page.

> In most environments where both Angular and AngularJS are used to render the application, both frameworks are loaded in the initial bundle being sent to the client. This not only results in a larger bundle size, but also means if a user only stays on Angular-rendered pages, the AngularJS framework and application is still loaded, and running, even if it is never accessed, potentially impacting overall application performance. 

Reviewer: This paragraph seems to be mixing too different aspects: (a) loading the AngularJS parts lazily and (b) bootstrapping the AngularJS app lazily.
They are not the same thing and impact the app differently.

From a bundling perspective, this can be done with modern bundling tools, such as the Angular CLI. From a performance perspective, you should only load, bootstrap, and render the AngularJS application when needed. The steps below show you how to create a service that lazy loads and boostraps AngularJS, create a component for AngularJS content, and how to configure the Angular Router to load AngularJS routes only on specific URLs.

### Create a service to lazy load AngularJS

> As of Angular version 8, lazy loading code can be accomplished simply by using the dynamic import syntax `import('..')`. In your application, you create a new service that uses dynamic imports to lazy load AngularJS.

Reviewer: _import('..')_ --> _import('...')_

<code-example path="upgrade-lazy-load-ajs/src/app/lazy-loader.service.ts" header="src/app/lazy-loader.service.ts">
</code-example>

> The service uses the `import()` method to load your bundled AngularJS application lazily. This decreases the initial bundle size of your application as you're not loading code your user doesn't need yet. You also need to provide a way to _bootstrap_ the application manually after it has been loaded. AngularJS provides a way to manually bootstrap an application using the `angular.bootstrap()` method with a provided HTML element. Your AngularJS app should also expose a `bootstrap` method that bootstraps the AngularJS app.

Reviewer: It would be useful to make `angular.bootstrap()` a link to the AngularJS docs (https://docs.angularjs.org/api/ng/function/angular.bootstrap).

<code-example path="upgrade-lazy-load-ajs/src/app/angularjs-app/index.ts" header="angularjs-app">
</code-example>

Your AngularJS application is configured with only the routes it needs to render content. The remaining routes in your application are handled by the Angular Router. The exposed `bootstrap` method is called in your Angular app to bootstrap the AngularJS application after the bundle is loaded.

<div class="alert is-important">

> **Note:** After AngularJS is loaded and bootstrapped, listeners such as those wired up in your route configuration will continue to listen for route changes. To ensure listeners are shut down when AngularJS isn’t being displayed, configure an `otherwise` option with the `$routeProvider` that renders an empty template. This assumes all other routes will be handled by Angular.

Reviewer: > isn’t

Fancy quote 🙈 

##
Also, a [link to `$routeProvider][1] would be useful.

[1]: https://docs.angularjs.org/api/ngRoute/provider/$routeProvider

</div>

### Create a component to render AngularJS content

In your Angular application, you need a component as a placeholder for your AngularJS content. This component uses the service you create to load and bootstrap your AngularJS app after the component is initialized.

<code-example path="upgrade-lazy-load-ajs/src/app/angular-js/angular-js.component.ts" header="src/app/angular-js/angular-js.component.ts">
</code-example>

When the Angular Router matches a route that uses AngularJS, the `AngularJSComponent` is rendered, and the content is rendered within the AngularJS [`ng-view`](https://docs.angularjs.org/api/ngRoute/directive/ngView) directive. 

### Configure a custom route matcher for AngularJS routes

> To configure the Angular router, you must define a route for AngularJS URLs. To match those URLs, you add a route configuration that uses the `matcher` property. The `matcher` allows you to use custom pattern matching for URL paths. The Angular router tries to match on more specific routes such as static and variable routes first. When it doesn’t find a match, it then looks at custom matchers defined in your route configuration. If the custom matchers don't match a route, it then goes to catch-all routes, such as a 404 page.

Reviewer: > doesn’t

Fancy quote 🙈

Define a custom matcher function for AngularJS routes:

> 

Reviewer 2: The following example defines a custom matcher function for AngularJS routes.

<code-example path="upgrade-lazy-load-ajs/src/app/app-routing.module.ts" header="src/app/app-routing.module.ts" region="matcher">
</code-example>

Then add a route object to your routing configuration using the `matcher` property and custom matcher, and the `component` property with `AngularJSComponent`.

> 

Reviewer 2: The following code adds a route object ...

<code-example path="upgrade-module/src/app/a-to-ajs-providers/hero-detail.component.ts" header="hero-detail.component.ts">
</code-example>

## Lazy Loading AngularJS

When building applications, you want to ensure that only the required resources are loaded when necessary. Whether that be loading of assets or code, making sure everything that can be deferred until needed keeps your application running efficiently. This is especially true when running different frameworks in the same application. When migrating large applications from AngularJS to Angular using a hybrid approach, you want to migrate some of the most commonly used features first, and only use the less commonly used features if needed. Doing so helps you ensure that the application is still providing a seamless experience for your users while you are migrating.

> 

Reviewer 2: [Lazy loading](guide/glossary#lazy-loading) is a technique that defers the loading of required assets and code resources until they are actually used. This reduces startup time and increases efficiency, especially when running different frameworks in the same application. 

When migrating...

In most environments where both Angular and AngularJS are used to render the application, both frameworks are loaded in the initial bundle being sent to the client. This results in both increased bundle size and possible reduced performance.

> Overall application performance is affected in cases where the user stays on Angular-rendered pages, because the AngularJS framework and application is still loaded, and running, even if it is never accessed.

Reviewer 3: ```suggestion
Overall application performance is affected in cases where the user stays on Angular-rendered pages because the AngularJS framework and application are still loaded and running, even if they are never accessed.
```

> To address the performance problem, you can use  [lazy loading](guide/glossary#lazy-loading) to only load, bootstrap, and render the AngularJS application when needed. This has the added benefit of reducing the initial bundle size.

Reviewer 3: ```suggestion
To address the performance problem, you can use  [lazy loading](guide/glossary#lazy-loading) to load, bootstrap, and render the AngularJS application only when needed. This has the added benefit of reducing the initial bundle size.
```

> The steps below steps below show you how to create a service that lazy loads and boostraps AngularJS, create a component for AngularJS content, and how to configure the Angular Router to load AngularJS routes only on specific URLs.

Reviewer 3: ```suggestion
The steps below show you how to create a service that lazy loads and bootstraps AngularJS, create a component for AngularJS content, and configure the Angular Router to load AngularJS routes only on specific URLs.
```

> 

Reviewer 2: You can take steps to mitigate both bundle size and performance issues. By isolating your AngularJS app to a separate bundle, you can take advantage of lazy loading to load, bootstrap, and render the AngularJS application only when needed. This reduces your initial bundle size, defers any potential impact from loading both frameworks until absolutely necessary, and keeps your application running as efficiently as possible.

> 

Reviewer 2: Make this a bullet list: 

The step below show you how to do the following:

* Set up a callback function for your AngularJS bundle.
* Create a service that lazy loads and bootstraps your AngularJS app.
* Create a routable component for AngularJS content.
* Create a custom `matcher` function for AngularJS-specific URLs and configure the Angular `Router` with the custom matcher for AngularJS routes.

<div class="alert is-important">

**Note:** After AngularJS is loaded and bootstrapped, listeners such as those wired up in your route configuration will continue to listen for route changes. To ensure listeners are shut down when AngularJS isn't being displayed, configure an `otherwise` option with the [$routeProvider](https://docs.angularjs.org/api/ngRoute/provider/$routeProvider) that renders an empty template. This assumes all other routes will be handled by Angular.

</div>

### Create a component to render AngularJS content

In your Angular application, you need a component as a placeholder for your AngularJS content. This component uses the service you create to load and bootstrap your AngularJS app after the component is initialized.

<code-example path="upgrade-lazy-load-ajs/src/app/angular-js/angular-js.component.ts" header="src/app/angular-js/angular-js.component.ts">
</code-example>

When the Angular Router matches a route that uses AngularJS, the `AngularJSComponent` is rendered, and the content is rendered within the AngularJS [`ng-view`](https://docs.angularjs.org/api/ngRoute/directive/ngView) directive. 

### Configure a custom route matcher for AngularJS routes

> To configure the Angular router, you must define a route for AngularJS URLs. To match those URLs, you add a route configuration that uses the `matcher` property. The `matcher` allows you to use custom pattern matching for URL paths. The Angular router tries to match on more specific routes such as static and variable routes first. When it doesn't find a match, it then looks at custom matchers defined in your route configuration. If the custom matchers don't match a route, it then goes to catch-all routes, such as a 404 page.

Reviewer 3: Should it always be Angular Router (capitalized)?

```suggestion
To configure the Angular Router, you must define a route for AngularJS URLs. To match those URLs, you add a route configuration that uses the `matcher` property. The `matcher` allows you to use custom pattern matching for URL paths. The Angular Router tries to match on more specific routes such as static and variable routes first. When it doesn't find a match, it then looks at custom matchers defined in your route configuration. If the custom matchers don't match a route, it then goes to catch-all routes, such as a 404 page.
```

Reviewer 2: It should only be capitalized (and put it code font) when referring to the API element. When used descriptively like this, it should remain lowercase.
